pkg/bindings/images/build.go

```diff
@@ -200,52 +228,116 @@ func Build(ctx context.Context, containerFiles []string, options entities.BuildO
 	}
 }
 
-func nTar(sources ...string) (io.ReadCloser, error) {
+func nTar(excludes []string, sources ...string) (io.ReadCloser, error) {
+	pm, err := fileutils.NewPatternMatcher(excludes)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error processing excludes list %v", excludes)
+	}
+
 	if len(sources) == 0 {
 		return nil, errors.New("No source(s) provided for build")
 	}
 
 	pr, pw := io.Pipe()
-	tw := tar.NewWriter(pw)
+	gw := gzip.NewWriter(pw)
+	tw := tar.NewWriter(gw)
 
 	var merr error
 	go func() {
 		defer pw.Close()
+		defer gw.Close()
 		defer tw.Close()
 
 		for _, src := range sources {
-			s := src
-			err := filepath.Walk(s, func(path string, info os.FileInfo, err error) error {
+			s, err := filepath.Abs(src)
+			if err != nil {
+				logrus.Errorf("cannot stat one of source context: %v", err)
+				merr = multierror.Append(merr, err)
+				return
+			}
+
+			err = filepath.Walk(s, func(path string, info os.FileInfo, err error) error {
 				if err != nil {
 					return err
 				}
-				if !info.Mode().IsRegular() || path == s {
-					return nil
-				}
 
-				f, lerr := os.Open(path)
-				if lerr != nil {
-					return lerr
+				if path == s {
+					return nil // skip root dir
 				}
 
 				name := strings.TrimPrefix(path, s+string(filepath.Separator))
-				hdr, lerr := tar.FileInfoHeader(info, name)
-				if lerr != nil {
-					f.Close()
-					return lerr
+
+				excluded, err := pm.Matches(filepath.ToSlash(name)) // nolint:staticcheck
+				if err != nil {
+					return errors.Wrapf(err, "error checking if %q is excluded", name)
 				}
-				hdr.Name = name
-				if lerr := tw.WriteHeader(hdr); lerr != nil {
-					f.Close()
-					return lerr
+				if excluded {
+					return nil
 				}
 
-				_, cerr := io.Copy(tw, f)
-				f.Close()
-				return cerr
+				if info.Mode().IsRegular() { // add file item
+					f, lerr := os.Open(path)
+					if lerr != nil {
+						return lerr
+					}
+
+					hdr, lerr := tar.FileInfoHeader(info, name)
+					if lerr != nil {
+						f.Close()
+						return lerr
+					}
+					hdr.Name = name
+					if lerr := tw.WriteHeader(hdr); lerr != nil {
+						f.Close()
+						return lerr
+					}
+
+					_, cerr := io.Copy(tw, f)
+					f.Close()
+					return cerr
+				} else if info.Mode().IsDir() { // add folders
+					hdr, lerr := tar.FileInfoHeader(info, name)
+					if lerr != nil {
+						return lerr
+					}
+					hdr.Name = name
+					if lerr := tw.WriteHeader(hdr); lerr != nil {
+						return lerr
+					}
+				} else if info.Mode()&os.ModeSymlink != 0 { // add symlinks as it, not content
+					link, err := os.Readlink(path)
+					if err != nil {
+						return err
+					}
+					hdr, lerr := tar.FileInfoHeader(info, link)
+					if lerr != nil {
+						return lerr
+					}
+					hdr.Name = name
+					if lerr := tw.WriteHeader(hdr); lerr != nil {
+						return lerr
+					}
+				} //skip other than file,folder and symlinks
+				return nil
```

Comment:
We shouldn't allow for symlinks that point to files/directory higher than the context directory.  i.e. if the context directory is `/home/user/runpodman` than pointing to `/home/user` should be invalid and I'm not seeing checks for this?

Reply:
This is more interesting to. I also asked for that too. However no proper answer. podman's local build does allow symlinks outside the contextdir and only copies symkink's itself. However containers/storage untar method checks if symlink target is outsided or not. hence i didnot put a check here. 

Also docker build allows this kind of symlinks. podman's build command should be identical to docker's build or not?

The problem is like that. When developer intentionally  creates symlinks such as targeted /etc /var etc? 

Local podman build's allow this however remote podman not. 

Previously I asked that how we handle this?

Reply:
@kazimsarikaya If there's a check within c/storage already, then I'm fine with that and you don't need to add one here.

Others may have different opinions, but from what I've been told, symlinks outside of the context-directory can cause potential security issues.  We do indeed try to match Docker's functionality, but not if it creates a security hole in the process.

Reply:
@TomSweeneyRedHat yes, if i send symlink outside the context-dir podman which runs at remote host returns exception because of https://github.com/containers/storage/blob/master/pkg/archive/archive.go#L668-L672 and build fails.

Reply:
The point is about remote attack.  If I put a symbolic link into the copy command and the path points to /etc/shadow on the host, I don't want that added to the container image.  So I think on the podman-remote build we need to make sure that links in the  context directory stay within the directory.  Are you saying that Docker allows you to send across a sympbolic link to the server that points to the hosts /etc/shadow, and can embed it into the container image?

Reply:
![Screen Shot 2021-01-14 at 00 54 41](https://user-images.githubusercontent.com/1861031/104515048-4bb55e00-5603-11eb-9406-0b5db88f05ed.png)


For symlinks you only create a symlink, not copy the content. Hence for your example, the shadow will be linked to the /etc/shadow inside container, not the content of build host.

Reply:
Ok SGTM

Reply:
However problem is at https://github.com/containers/storage/blob/master/pkg/archive/archive.go#L668-L672 hence this problem isnot related to this pr. hence it seems this conversation is resolved.